core/src/cuda_runtime.rs

```diff
@@ -226,6 +238,34 @@ impl<T: Clone> PinnedVec<T> {
         self.check_ptr(old_ptr, old_capacity, "resize");
     }
 
+    pub fn append(&mut self, other: &mut Vec<T>) {
+        let old_ptr = self.x.as_mut_ptr();
+        let old_capacity = self.x.capacity();
+        // Predict realloc and unpin.
+        if self.pinned && self.x.capacity() < self.x.len() + other.len() {
+            unpin(old_ptr);
+            self.pinned = false;
+        }
```

Comment:
nits:
This code can be shared with a private helper `(old_ptr, old_capacity) = fn prepare_to_change`, because it is same as [this](https://github.com/solana-labs/solana/pull/6604/files#diff-a2c89f2a8e8cd92be22e8afd760a89e1R253) and (untouched by this PR) `resize`, `push`.

Ideally, could such a helper `fn` be made to take a closure, resulting like this?: `fn with_prepared_to_change(other, need_unpin, || {self.x.{append/push/resize}()})` with no runtime cost with Rust?

Personally, I have relatively low tolerance for duplicate code. :p

Reply:
Yea. I agree. I'm not sure of the cost of the closure, but I can check.

Reply:
actually the condition is different except for append and append_pinned